### PR TITLE
feature(provisioner) Add a default mssql provisioner

### PR DESCRIPTION
**Pull request by Hentouane**
Adding a new default provider for a mssql database.

#### Description
Created a new default provider template for a mssql db based on already existing mysql and postgres ones. Updated zz-default.provisioners.yaml and README.md.

#### What does this PR do?
Adding a new default provider for a mssql database per feature request https://github.com/score-spec/score-k8s/issues/66

Was tested with the following score.yaml file:

```
apiVersion: score.dev/v1b1
metadata:
    name: test-container
containers:
    main:
        image: mcr.microsoft.com/mssql-tools
        variables:
            SERVER_NAME: ${resources.db.host}
            MSSQL_SA_USER: ${resources.db.username}
            MSSQL_SA_PASSWORD: ${resources.db.password}
        command:
            - /bin/sh
            - -c
            - /opt/mssql-tools/bin/sqlcmd -S $SERVER_NAME -U $MSSQL_SA_USER -P $MSSQL_SA_PASSWORD -Q "select @@version"
resources:
    db:
        type: mssql
```

Once deployed in a k3s cluster, the following was succesfully logged by the test-container:

```
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
Microsoft SQL Server 2022 (RTM-CU15-GDR) (KB5046059) - 16.0.4150.1 (X64)                                                                                                                                      
     Sep 25 2024 17:34:41                                                                                                                                                                                      
     Copyright (C) 2022 Microsoft Corporation                                                                                                                                                                  
     Developer Edition (64-bit) on Linux (Ubuntu 22.04.5 LTS) <X64>                                                                                                                                            
                                                                                                                                                                                                               
  (1 rows affected)
```


#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New chore (expected functionality to be implemented)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I've signed off with an email address that matches the commit author.

### CODE REVIEW

File: internal/provisioners/default/zz-default.provisioners.yaml [hunk 1]
```diff
@@ -798,3 +798,120 @@
...
+  type: mssql
+  init: |
+    randomDatabase: db-{{ randAlpha 8 }}
+    randomPassword: {{ randAlphaNum 16 | quote }}
+  state: |
+    service: mssql-{{ .SourceWorkload }}-{{ substr 0 8 .Guid | lower }}
+    database: {{ dig "database" .Init.randomDatabase .State | quote }}
+    username: sa
+    password: {{ dig "password" .Init.randomPassword .State | quote }}
+  outputs: |
```
Comment (mathieu-benoit) [inline internal/provisioners/default/zz-default.provisioners.yaml:812]: I understand why we have 4 distinct `outputs` here given the `sqlcmd` example taking separate parameters.

But please could you also add a more generic one `connection`, in this case that will return:
```
connection: "Server=tcp:{{ .State.service }},1433;Initial Catalog={{ .State.database }};User="
```

There are multiple benefits with this `connection` field:
- Abstracting from the Devs about how to build the connection string
- There are more fields in other scenario for example with Azure SQL: `"Server=tcp:server.fully_qualified_domain_name,1433;Initial Catalog=database_name;Encrypt=True;TrustServerCertificate=False;Connection Timeout=30;Authentication=Active Directory Default";User ID=administrator_login;Password=administrator_login_password"`
- As Dev, you don't have to change your Score file if you point to an in-cluster database or a managed database, etc.

**Review by mathieu-benoit — CHANGES_REQUESTED** [review]
Thanks @Hentouane for your contribution.

I have left a few comments for updates.

Also, you need to sign your commits to be compliant with the DCO check, see here how to resolve it: https://github.com/score-spec/score-k8s/runs/31986670494.

File: internal/provisioners/default/zz-default.provisioners.yaml [hunk 2]
```diff
@@ -798,3 +798,120 @@
...
+    host: {{ .State.service }}
```
[the remark this answers is not in the document]
Comment (Hentouane) [reply internal/provisioners/default/zz-default.provisioners.yaml]: Makes sense to me, variable was renamed.

File: internal/provisioners/default/zz-default.provisioners.yaml [hunk 3]
```diff
@@ -798,3 +798,120 @@
...
+    port: 1433
+    name: {{ .State.database }}
```
[the remark this answers is not in the document]
Comment (Hentouane) [reply internal/provisioners/default/zz-default.provisioners.yaml]: Variable was renamed though I hardcoded it to the 'master' database, as there seems to not have an equivalent to MySQL's 'MYSQL_DATABASE' environement variable for MSSQL.

Source: https://learn.microsoft.com/en-us/sql/linux/sql-server-linux-configure-environment-variables?view=sql-server-ver16#environment-variables

File: internal/provisioners/default/zz-default.provisioners.yaml (hunk 1 above)
Comment (Hentouane) [reply internal/provisioners/default/zz-default.provisioners.yaml:811]: Connection output variable was added

**Comment by Hentouane** [comment]
Commits were signed-off and changes were applied. Leaving one conversation open as to confirm reviewer's approval.

**Review by mathieu-benoit — APPROVED** [review]

**Comment by mathieu-benoit** [comment]
Thanks, @Hentouane!

LGTM

@astromechza, another pair of eyes for this review please? Thanks1

**Review by astromechza — APPROVED** [review]
Yup, this looks good to me! Definitely a good starting point for those using mssql

Thanks @Hentouane 👍

**Pull request merged**